src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -0,0 +1,139 @@
+package is.hail.expr.ir.functions
+
+import is.hail.asm4s._
+import is.hail.expr.{AST, FunType, Lambda, TypeTag}
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.utils._
+import is.hail.asm4s.coerce
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object IRFunctionRegistry {
+
+  val registry: mutable.Map[String, Seq[(Seq[Type], Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  def addIRFunction(f: IRFunction) {
+    val l = registry(f.name)
+    registry.put(f.name,
+      l :+ (f.argTypes, { args: Seq[IR] =>
+        ApplyFunction(f, args)
+      }))
+  }
+
+  def addIR(name: String, types: Seq[Type], f: Seq[IR] => IR) {
+    val l = registry(name)
+    registry.put(name, l :+ ((types, f)))
+  }
+
+  def lookupFunction(name: String, args: Seq[Type]): Option[Seq[IR] => IR] = {
+    val validMethods = registry(name).flatMap { case (ts, f) =>
+      if (ts.length == args.length) {
+        ts.foreach(_.clear())
+        if ((ts, args).zipped.forall(_.unify(_)))
+          Some(f)
+        else
+          None
+      } else {
+        None
+      }
+    }
+    validMethods match {
+      case Seq() => None
+      case Seq(x) => Some(x)
+      case _ => fatal(s"Multiple methods found that satisfy $name$args.")
+    }
+  }
+
+  UtilFunctions.registerAll()
+}
+
+abstract class RegistryFunctions {
+
+  def registerAll(): Unit
+
+  private val tvBoxes = mutable.Map[String, Box[Type]]()
+
+  def tv(name: String): TVariable =
+    TVariable(name, b=tvBoxes.getOrElseUpdate(name, Box[Type]()))
```

Comment:
spaces around =.  I'd add a (private) box lookup convenience function since you use this expression twice.